godot-core/src/builtin/collections/packed_array.rs

```diff
@@ -378,6 +378,18 @@ macro_rules! impl_packed_array {
             pub fn as_inner(&self) -> inner::$Inner<'_> {
                 inner::$Inner::from_outer(self)
             }
+
+            /// Moves elements from slice, overwriting any data from packed array.
+            ///
+            /// # Safety
+            ///
+            /// * Pointer must be valid and not points to `self` data.
+            /// * Length must be equal.
+            /// * All data in `self` can be safely overwritten. At worst, it leaks the old value.
+            ///   But default-constructed [`GString`] is a null pointer anyway, so no need to drop it.
```

Comment:
I don't think this is a very strong reasoning. If there are elements, every old element should be dropped properly.

Alternatively, add a `debug_assert` that verifies that this is not the case. This should happen for each precondition where it's possible.

Reply:
I added a test case `packed_array_resize_zeros_str` to check for that invariance. Other element types don't have `Drop` impls.

Also, what is Godot possibly do with null value anyways? And leaks are still memory-safe.

Reply:
> Also, what is Godot possibly do with null value anyways? And leaks are still memory-safe.
> 
> Edit: I don't mean leaks are _okay_. But with previous test it should not leak, as null pointer don't point to any allocation.

That doesn't matter -- if we ever add `Drop` logic in any of the builtin types, we will 100% not think of updating the `From` logic inside `PackedArray`. In such cases, we can then easily introduce memory leaks, or worse, logic errors that propagate i-don't-know-how far.

Each element must be dropped, no matter how it was initialized.

By the way, you cannot test for leaks in integration test. You need a memory checker like LSan, and even that is not always entirely accurate. You cannot check for other logic errors due to missing `Drop` in a general way, either. Let's just implement this correctly :wink:

Reply:
Okay fine. I'll `drop_in_place` all elements first then `copy_nonoverlapping`. There's minor risk that if any of the `Drop` panics we're in a pickle, but from slice has also that risk too.

Reply:
Good point. `Drop` panicking is often an issue, but I think it's fair to assume this doesn't happen for Godot builtin types (and if it does, it would due to some violated postcondition on our side, in which case abort would be OK).

But maybe mention this in a comment?